File: app/src/main/java/net/bloople/manga/Book.java
package net.bloople.manga;

import com.bumptech.glide.load.model.GlideUrl;
import com.bumptech.glide.load.model.LazyHeaderFactory;
import com.bumptech.glide.load.model.LazyHeaders;
import com.dslplatform.json.CompiledJson;

import java.util.ArrayList;
import java.util.List;

import okhttp3.Credentials;

@CompiledJson(onUnknown = CompiledJson.Behavior.FAIL)
public class Book {
    public String path;
    public String pagePaths;
    public int pages; //We are trusting the server that pages matches pagePathsList().size()
    public int publishedOn;
    public String key;
    public List<String> tags;

    private Library library;
    private String title;
    private String normalisedTitle;
    private long _id;
    private ArrayList<String> pagePathsList;

    public Library library() {
        return library;
    }

    public MangosUrl thumbnailUrl() {
        return library.mangos().withAppendedPath("/img/thumbnails/" + key + ".jpg");
    }

    public String title() {
        return title;
    }

    String normalisedTitle() {
        return normalisedTitle;
    }

    public long id() {
        return _id;
    }

    private ArrayList<String> pagePaths() {
        if(pagePathsList == null) pagePathsList = new PagesInflater(pagePaths).inflate();
        return pagePathsList;
    }

    MangosUrl pageUrl(int index) {
        return library.rootUrl().withAppendedPath("/" + path + "/" + pagePaths().get(index));
    }

    void inflate(Library library) {
        this.library = library;
        _id = Long.parseLong(key.substring(0, 15), 16); //Using substring of key would be dangerous for large N
        normalisedTitle = path.replaceAll("[^A-Za-z0-9]+", "").toLowerCase();
        title = path.replaceAll("\\s+", " ");
        this.library.books().put(_id, this);
    }
}

File: app/src/main/java/net/bloople/manga/Book.kt
package net.bloople.manga

import kotlinx.serialization.SerialName
import java.util.ArrayList

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

@Serializable
class Book(
    val path: String,
    @SerialName("pagePaths")
    val pagePathsDeflated: String,
    val pages: Int,
    val publishedOn: Int,
    val key: String,
    val tags: List<String>
) {
    @Transient
    lateinit var library: Library
    val title: String by lazy { path.replace("\\s+".toRegex(), " ") }
    val normalisedTitle: String by lazy { path.replace("[^A-Za-z0-9]+".toRegex(), "").lowercase() }
    val id: Long by lazy { key.substring(0, 15).toLong(16) } //Using substring of key would be dangerous for large N
    private val pagePaths: ArrayList<String> by lazy { inflatePagePaths(pagePathsDeflated) }
    val thumbnailUrl: MangosUrl by lazy { library.thumbnailsUrl / "$key.jpg" }

    fun pageUrl(index: Int): MangosUrl {
        return library.rootUrl / path / pagePaths[index]
    }
}
